### PR TITLE
Call ensure_options methods before initialization

### PR DESCRIPTION
**Pull request by leogermani**
#### Changes proposed in this Pull Request:

* Make Config call the `ensure_options_$feature` methods before the initialization of the packages

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

p1660822904458439-slack-C02NENKA4GN

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

At this point, nothing should change, so test for regressions.

Let's test Sync to see if the behavior was kept the same.

* Actiave only VideoPress plugin
* Go to Jetpack Debug > Sync Data Settings Tester and checks the result
* Activate Protect and check the results again. `active_plugins` and `stylesheet` should have been added to the list of options. 

Run the test above on trunk and with this branch checked out. The results must be the same.

Check the docs to see if they are clear and correct.

### CODE REVIEW

**Comment by oskosk** [comment]
Just pushed a rebase in https://github.com/Automattic/jetpack/pull/25710/commits/435442fe0f35dd04b41d8b44ab9d95fcc6120c97

File: projects/packages/config/README.md
Comment (zinigor) [inline projects/packages/config/README.md:121]: This is a great start, but what do you think about passing the options directly into the called method? This will let us avoid the back and forth.

The downside of that is that we're passing the burden on to the package, but we are already doing it for package initialization code.

One other thing that I have in mind is that we don't overwrite feature options when we ensure settings, but rather merge the passed arrays, and store that in Config properties. But then there's a caveat - we are not using the same Config instance everywhere, we're just creating a new one.

What if we did it like a filter with a closure? Sort of like

```
$options = apply_filters( "jetpack_config_$feature_options", $default_options );
```

And then in the Config `ensure` method we say:

```
public function ensure_options_configurable_package( $options ) {
    add_filter( 'jetpack_config_configurable_package_options', function( $defaults ) use ( $options ) {
        return array_merge ( $defaults, $options );
    } );
```

**Review by zinigor — CHANGES_REQUESTED** [review]
I have replied in the code with a suggestion, can you please take a look?

File: projects/packages/config/README.md
Comment (leogermani) [reply projects/packages/config/README.md:121]: > The downside of that is that we're passing the burden on to the package, but we are already doing it for package initialization code.

I think that's the best thing to do. Let the packages themselves decide what to do.

> What if we did it like a filter with a closure? Sort of like

I don't think Config should merge the settings. We should let packages handle them. A couple of examples why I think that:

If you look at what Connection does for example, it does not merge the options. It adds a new `Plugin` for each consumer that initializes it. It uses this rather as a hook to create a list of all consumers that are using the Connection.

Also, how we are going to merge the options? For example, if a consumer says that options A is `true` and another consumer says that option B is `false`.

Only the package will know what value should "win", depending on its business logic.

For the VideoPress package, for example, if any consumer sets `admin_ui` as `true`, I want it to override any other places where it might be set to `false`. This is a case by case decision that only the package knows how to handle.

> This is a great start, but what do you think about passing the options directly into the called method?

We could do that and avoid one function call... but it's good to store the options in the instance as we do now so it's also available in the `enable_$feature` method

Comment (leogermani) [reply projects/packages/config/README.md:121]: if you look at how Sync merge the options for example, it is not straight forward... many specific rules there...

Comment (zinigor) [reply projects/packages/config/README.md:121]: OK, so I think I misunderstood your intention back there. Now I see you would like run the options initialization at the time of feature declaration, not at the time the feature gets "ensured" at the `plugins_loaded` time.

I agree with you that Config shouldn't merge options, and that more complicated logic is sometimes involved in deciding what option "wins". Config was initially considered an "opt-in" mechanism, not really an "opt-in and configure", which it is now becoming. Let's try to go this route and see where it comes.

This does let us to move the custom Config code into packages, so maybe we'll look at that later, so that the logic feels more at home in the package maintainer's hands.

All that to say - cool, thanks for doing this, let's try to go this route.

**Review by zinigor — APPROVED** [review]

**Pull request merged**
